File: 101-SymmetricTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
// Iterative Solution
class Solution {
    public:
        bool isSymmetric(TreeNode* root){
            if(!root) return true;
            stack<TreeNode*> s;
            s.emplace(root->left);
            s.emplace(root->right);
            
            while(!s.empty()) {
                TreeNode* t1 = s.top();
                s.pop();
                TreeNode* t2 = s.top();
                s.pop();
                if (!t1 && !t2) continue;
                if (!t1 || !t2 || t1->val != t2->val) return false;
                s.emplace(t1->left);
                s.emplace(t2->right);
                s.emplace(t1->right);
                s.emplace(t2->left);
            }
            return true;
        }
};
// Recursive solution 
class Solution2 {
public:
    bool isSymmetric(TreeNode* root) {
        if (!root) {
            return true;
        }
        return isSymmetricHelper(root->left, root->right);
    }
    
    bool isSymmetricHelper(TreeNode *left, TreeNode *right) {
        if (!left && !right) {
            return true;
        }
        if (!left || !right || left->val != right->val) {
            return false;
        }
        return isSymmetricHelper(left->left, right->right) &&
               isSymmetricHelper(left->right, right->left);
    }
};

File: java/101-SymmetricTree.java
/*

Given a binary tree, check whether it is a mirror of itself (ie, symmetric around its center).

For example, this binary tree [1,2,2,3,4,4,3] is symmetric:

    1
   / \
  2   2
 / \ / \
3  4 4  3
But the following [1,2,2,null,3,null,3] is not:
    1
   / \
  2   2
   \   \
   3    3
Note:
Bonus points if you could solve it both recursively and iteratively.

*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        return helper(root, root);
    }
    private boolean helper(TreeNode p, TreeNode q) {
        if(p==null && q==null) return true;
        if((p==null && q!=null) || (p!=null && q==null) || (p.val != q.val)) return false;
        return helper(p.left, q.right) && helper(p.right, q.left);
    }
}
